_posts/2018-01-15-contributing-to-prettier-atom.md

```diff
@@ -0,0 +1,219 @@
+
+In other words, it's a [flame graph](http://www.brendangregg.com/FlameGraphs/cpuflamegraphs.html).
+
+By hovering over these bars, I was able to find the one corresponding to the `prettier-atom.activate` function and understand what was causing the slow activation time.
+
+Looking at the chart, a few things stood out.
+
+* Loading code adds up. Every `require` triggers a file to be read, parsed and compiled.
+* One file was requiring something called `toConsumableArray` which was taking 64ms to load.
+* Another file was requiring `lodash/fp` which was taking 111ms to run.
+
+![consumable array load time](/assets/posts/consumable-array-load-time.png)
+![lodash load time](/assets/posts/lodash-load-time.png)
+
+These were both occurring from a `require`. That means it wasn't time spent computing useful stuff, it was time spent simply loading the libraries. Yuck!
+
+If we could avoid loading these packages, we would save ~170ms from the activation time!
+
+## Optimization 1: Removing toConsumableArray
+
+I noticed that the `toConsumableArray` import was coming from `babel-runtime`. Babel runtime is a collection of polyfills to allow ES6 syntax in ES5 code. I looked at the file triggering the require, and sure enough there was some fancy syntax going on.
+
+```js
+const getAllScopes = () => [
+	...getJavascriptScopes(),
+	...getTypescriptScopes(),
+	...getCssScopes(),
+	...getJsonScopes(),
+	...getGraphQlScopes(),
+	...getMarkdownScopes()
+];
+```
+
+The `...` operator allows the arrays to be cleanly combined, but it's also not supported in ES5. So babel sees that syntax and adds `toConsumableArray` to polyfill the functionality.
+
+It's nice syntax, but not nice enough to warrant 64ms of activation time. Converting the expression to an ES5 compatible version stopped `babel-runtime` from being included.
+
+```js
+const getAllScopes = () =>
+	[
+		getJavascriptScopes(),
+		getTypescriptScopes(),
+		getCssScopes(),
+		getJsonScopes(),
+		getGraphQlScopes(),
+		getMarkdownScopes()
+	].reduce((allScopes, scopes) => allScopes.concat(scopes), []);
+```
+
+Sure enough, the require disappeared and the activation time reduced substantially.
+
+## Optimization 2: Lazy loading lodash
+
+A common pattern in nodejs code is to include all your requires at the top of the file.
+
+Here's an example:
+
+```js
+const _ = require("lodash");
+
+function reverseAndCompact(arr) {
+	const reversed = _.reverse(arr);
+	return _.compact(reversed);
+}
+
+function activate() {
+	console.log("Activating!");
+}
+```
+
+This is nice for keeping track of what's required, but slows down the initial load of the file. Anything that requires the above file will need to wait for `lodash` to load before it can even see the `reverseAndCompact` function.
+
+This is especially annoying if we want to use the `activate` function. It doesn't depend on `lodash`, but still needs to wait for `lodash` if we import it.
```

Comment:
Super minor, but my brain hiccuped a bit: I was expecting "This is especially annoying if we JUST  want to use the `activate` function", or something like that. Obviously not a blocker :)